src/hotspot/cpu/ppc/stubGenerator_ppc.cpp

```diff
@@ -3544,6 +3544,461 @@ class StubGenerator: public StubCodeGenerator {
+                     (unsigned char)0b11111000,
+      /* 10       */ (unsigned char)0b11110000,
+      /* 11       */ (unsigned char)0b01010100,
+      /* 12 .. 14 */ (unsigned char)0b01010000, (unsigned char)0b01010000, (unsigned char)0b01010000,
+      /* 15       */ (unsigned char)0b01010100 ) };
+
+    static const unsigned char VEC_ALIGN maskLUT_URL_val[16] = {
+      ARRAY_TO_LXV_ORDER(
+      /* 0        */ (unsigned char)0b10101000,
+      /* 1 .. 9   */ (unsigned char)0b11111000, (unsigned char)0b11111000, (unsigned char)0b11111000, (unsigned char)0b11111000,
+                     (unsigned char)0b11111000, (unsigned char)0b11111000, (unsigned char)0b11111000, (unsigned char)0b11111000,
+                     (unsigned char)0b11111000,
+      /* 10       */ (unsigned char)0b11110000,
+      /* 11 .. 12 */ (unsigned char)0b01010000, (unsigned char)0b01010000,
+      /* 13       */ (unsigned char)0b01010100,
+      /* 14       */ (unsigned char)0b01010000,
+      /* 15       */ (unsigned char)0b01110000 ) };
+
+    static const unsigned char VEC_ALIGN bitposLUT_val[16] = {
+      ARRAY_TO_LXV_ORDER(
+      0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, (unsigned char)0x80,
+      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 ) };
+
+    static const unsigned char VEC_ALIGN pack_lshift_val[16] = { 2, 4, 6, 0, 2, 4, 6, 0, 2, 4, 6, 0, 2, 4, 6, 0 };
+
+    static const unsigned char VEC_ALIGN pack_rshift_val[16] = { 0, 4, 2, 0, 0, 4, 2, 0, 0, 4, 2, 0, 0, 4, 2, 0 };
+
+    // The last 4 index values are "don't care" because
+    // we only use the first 12 bytes of the vector,
+    // which are decoded from 16 bytes of Base64 characters.
+    static const unsigned char VEC_ALIGN pack_permute_val[16] = {
+      14, 13, 12,
+      10,  9,  8,
+       6,  5,  4,
+       2,  1,  0,
+       0,  0,  0, 0 };
+
+    static const unsigned char VEC_ALIGN p10_pack_permute_val[16] = {
+      10, 11, 12, 13, 14, 15,
+       2,  3,  4,  5,  6,  7,
+       0,  0,  0,  0 };
+
+    // loop_unrolls needs to be a power of two so that the rounding can be
+    // done using a mask.
+    //
+    // The amount of loop unrolling was determined by running a benchmark
+    // that decodes a 20k block of Base64 data on a Power9 machine:
+    // loop_unrolls = 1 :
+    // (min, avg, max) = (101097.676, 101611.387, 101838.460), stdev = 182.156
+    // loop_unrolls = 2 :
+    // (min, avg, max) = (106430.268, 106717.050, 106977.072), stdev = 171.667
+    // loop_unrolls = 4 :
+    // (min, avg, max) = (105342.240, 106084.930, 106390.059), stdev = 264.617
+    // loop_unrolls = 8 :
+    // min, avg, max) = (105790.841, 106318.585, 106602.132), stdev = 243.740
+    //
+    // Given that the performance is about 5% better at 2, and that there's
+    // an advantage to keeping loop_unrolls small (to be able to process
+    // smaller buffers), 2 is clearly the best choice.
+    const unsigned loop_unrolls = 2;
```

Comment:
Unrolling should be re-evaluated after alignment is fixed. align(32) is currently at the wrong place (see my comment below).

Reply:
Agreed.